pkg/router/kubernetes_default.go

```diff
@@ -162,3 +163,69 @@ func (c *KubernetesDefaultRouter) reconcileService(canary *flaggerv1.Canary, nam
 
 	return nil
 }
+
+//Finalize reverts the apex router if not owned by the Flagger controller.
+func (c *KubernetesDefaultRouter) Finalize(canary *flaggerv1.Canary) error {
+	apexName, _, _ := canary.GetServiceNames()
+
+	svc, err := c.kubeClient.CoreV1().Services(canary.Namespace).Get(apexName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	//No need to do any reconciliation if the router is owned by the controller
+	if hasCanaryOwnerRef, isOwned := c.isOwnedByCanary(svc, canary.Name); !hasCanaryOwnerRef && !isOwned {
+		//If kubectl annotation is present that will be utilized, else reconcile
+		if a, ok := svc.Annotations[kubectlAnnotation]; ok {
+			var storedSvc corev1.Service
+			err := json.Unmarshal([]byte(a), &storedSvc)
+			if err != nil {
+				return fmt.Errorf("router %s.%s failed to unMarshal annotation %s, unable to revert",
+					svc.Name, svc.Namespace, kubectlAnnotation)
+			}
+			clone := svc.DeepCopy()
+			clone.Spec = storedSvc.Spec
+
+			//ignore immutable fields
+			if svc.Spec.ClusterIP != "" {
```

Comment:
Instead of dealing with immutable fields I think it would be better to restore only the selector, since Flagger changes only that.

Reply:
I like that idea, easier implementation.